bot/exts/evergreen/splitify.py

```diff
@@ -0,0 +1,130 @@
+import math
+import random
+from io import BytesIO
+
+import discord
+from PIL import Image
+from discord.ext import commands
+
+
+MAX_SQUARES = 193_600
+
+
+class Splitify(commands.Cog):
+    """Splits your avatar in x squares!"""
+
+    def __init__(self, bot: commands.Bot) -> None:
+        self.bot = bot
+
+    @staticmethod
+    def split_image(img: Image, squares: int) -> list:
+        """Splits the image into x squares."""
+        width, heigth = img.size
+
+        xy = math.sqrt(squares)
+
+        x_frac = width // xy
+        y_frac = heigth // xy
+
+        left, top, right, bottom, = 0, 0, x_frac, y_frac
+
+        new_imgs = []
+
+        for index in range(squares):
+            new_img = img.crop((left, top, right, bottom))
+            new_imgs.append(new_img)
+
+            if (index + 1) % xy == 0:
+                top += y_frac
+                bottom += y_frac
+                left = 0
+                right = x_frac
+            else:
+                left += x_frac
+                right += x_frac
+
+        return new_imgs
+
+    @staticmethod
+    def join_images(imgs: list) -> Image:
+        """Stitches all the image squares into a new image."""
```

Comment:
Could you explain the algorithm you used for solving this task in the doc-string?

Reply:
Is there an issue with it or are you just personally interested?

Reply:
Here it is.

1. I shuffle passed imgs. Reason: to randomize the squares (split pieces).

2. I get a single square (split piece) out of the list and define single_width as the square's width and single_height as the square's height.

3. I get the root of type integer of the number of imgs (split pieces) in the list and call it multiplier. I then proceed to calculate total height and width of the new image that I am creating using the same multiplier.

4. I then define new_img as the image that I am creating, using the previously obtained total_width and total_height.

5. Now I define width_multiplier as well as height with values of 0. These will be used to correctly position squares (split pieces) onto the new_img canvas.

6. Similar to how in the split_image function, I get the root of number of imgs (split pieces) in the list. In split_image function, it was the passed squares (number of squares) instad of a number of imgs in the list that I got the square of here.

7. In the for loop, as I iterate, I multiply single_width by width_multiplier to correctly position a square (split piece) width wise. I then proceed to paste the newly positioned square (split piece) onto the new_img. I increase the width_multiplier by 1 every iteration so the image wouldn't get pasted in the same spot and the positioning would move accordingly. I make sure to increase the width_multiplier before the check, which checks if the end of a row has been reached, - (index + 1) % pieces == 0, so after it, if it was True, width_multiplier would have been reset to 0 (start of the row). If the check returns True, the height gets increased by a single square's (split piece) height to lower the positiong height wise and, as I just mentioned, the width_multiplier gets reset to 0 and width will then be calculated from the start of the new row. The for loop finishes once all the squres were positioned accordingly.

8. Finally, I return the new_img, the randomized squares (split pieces) stitched back into the format of the original image - user's PFP.